### DISCUSSION TITLE
Create pixmap directly from PIL image?

### DISCUSSION THREAD

**Discussion by vs-777** [pymupdf/PyMuPDF#801] [Looking for help] [answered]
This is my problem: in one function, I generate an image using PIL. In another function, I get this image and create a pixmap so that I can insert it on a pdf. Currently, I have to save the PIL image into a folder, and load it when calling the Pixmap class. Is there anyway to avoid the saving process and directly load the PIL image data into a pymupdf pixmap? e.g.:

```
import fitz
from PIL import Image
im = Image.open("picture.jpg")
pix = fitz.Pixmap(im)
```

**Comment by JorjMcKie** [comment, marked as answer]
Several comments:
1. You do not need a pixmap for PDF image insertion: a file or a binary string of the image file or a ``io.BytesIO`` are accepted as well.
2. Similar is true for pixmaps themselves: among the many constructors use `pix = fitz.Pixmap(stuff)`. If "stuff" is a string, it is interpreted as a filename. If it is a `bytes`, `bytearray` or ``io.BytesIO`` object, the constructor assumes an image.

So if you have an PIL Image, do this to insert in a PDF page:
```python
import io
from PIL import Image

bio = io.BytesIO()
img.save(bio, "JPEG", ...)

# option 1: use io.BytesIO
page.insertImage(rect, stream=bio)

# option 2: use pixmap
pix = fitz.Pixmap(bio)
page.insertImage(rect, pixmap=pix)
```

The point is that PIL Images must be saved somehow first. This can happen to memory using io.BytesIO.

**Comment by vs-777** [comment]
Perfect! Didn't know about `io`. Thank you.
